goto-instrument 5.84 crash when trying to use __CPROVER_pointer_in_range_dfcc contract

I am using CBMC 5.84 on Apple M1 macOS.

I'm trying to verify a simple algorithms that loops over an array of data that is passed in as a parameter.
I think I need to use __CPROVER_pointer_in_range_dfcc to verify that my pointer(s) remain valid for the array of data that has been passed in to my function.

See https://github.com/rod-chapman/cbmc-examples/tree/main/arrays

In that directory, 
```
make TARGET=arsum
```
yields
```
rodchap@f4d4889dcf6d arrays % make TARGET=arsum
rm -f *.goto
goto-cc --function arsum_harness -DCBMC -o ar.goto ar.c
goto-instrument --bounds-check --pointer-check --memory-cleanup-check --div-by-zero-check --signed-overflow-check --pointer-overflow-check --conversion-check --undefined-shift-check --enum-range-check --pointer-primitive-check --dfcc arsum_harness --apply-loop-contracts --enforce-contract arsum ar.goto ar_contracts.goto
Reading GOTO program from 'ar.goto'
Function Pointer Removal
Virtual function removal
Cleaning inline assembler statements
Loading CPROVER C library (arm64)
Adding the cprover_contracts library (arm64)
file <builtin-library-malloc> line 6: symbol '__CPROVER_malloc_is_new_array' already has an initial value
symbol '__CPROVER_alloca_object' already has an initial value
symbol '__CPROVER_new_object' already has an initial value
file <builtin-library-free> line 11: symbol '__CPROVER_malloc_is_new_array' already has an initial value
Instrumenting memory predicate 'arsum'
Instrumenting memory predicate 'arsum_harness'
Instrumenting harness function 'arsum_harness'
Wrapping 'arsum' with contract 'arsum' in CHECK mode
--- begin invariant violation report ---
Invariant check failed
File: /tmp/cbmc-20230525-15049-1i7eyc/src/util/std_expr.cpp:173 function: instantiate
Condition: variables.size() == values.size()
Reason: Precondition
Backtrace:
0   goto-instrument                     0x0000000100b9bd4c _Z15print_backtraceRNSt3__113basic_ostreamIcNS_11char_traitsIcEEEE + 120
1   goto-instrument                     0x0000000100b9c09c _Z13get_backtracev + 44
2   goto-instrument                     0x000000010076a224 _Z29invariant_violated_structuredI17invariant_failedtJRKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEEENS1_9enable_ifIXsr3std10is_base_ofIS0_T_EE5valueEvE4typeES9_S9_iS9_DpOT0_ + 52
3   goto-instrument                     0x000000010076a1f0 _Z29invariant_violated_structuredI17invariant_failedtJRKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEEENS1_9enable_ifIXsr3std10is_base_ofIS0_T_EE5valueEvE4typeES9_S9_iS9_DpOT0_ + 0
4   goto-instrument                     0x0000000100c0d7d8 _ZNK13binding_exprt11instantiateERKNSt3__16vectorI5exprtNS0_9allocatorIS2_EEEE + 804
5   goto-instrument                     0x00000001007c4450 _ZN21dfcc_wrapper_programt23encode_requires_clausesEv + 220
6   goto-instrument                     0x00000001007c26c4 _ZN21dfcc_wrapper_programtC2E19dfcc_contract_modetRK7symboltS3_RK24dfcc_contract_functionstS3_R11goto_modeltR16message_handlertR13dfcc_librarytR16dfcc_instrumenttR28dfcc_lift_memory_predicatest + 1868
7   goto-instrument                     0x00000001007a214c _ZN22dfcc_contract_handlert25add_contract_instructionsE19dfcc_contract_modetRK8dstringtS3_S3_RK7symboltR13goto_programtRNSt3__13setIS1_NS9_4lessIS1_EENS9_9allocatorIS1_EEEE + 176
8   goto-instrument                     0x00000001007bdd00 _ZN19dfcc_swap_and_wrapt14check_contractE24dfcc_loop_contract_modetRK8dstringtS3_RNSt3__13setIS1_NS4_4lessIS1_EENS4_9allocatorIS1_EEEEb + 1604
9   goto-instrument                     0x00000001007bd3a0 _ZN19dfcc_swap_and_wrapt13swap_and_wrapE19dfcc_contract_modet24dfcc_loop_contract_modetRK8dstringtS4_RNSt3__13setIS2_NS5_4lessIS2_EENS5_9allocatorIS2_EEEEb + 148
10  goto-instrument                     0x0000000100798560 _ZN5dfcct21wrap_checked_functionEv + 180
11  goto-instrument                     0x0000000100796fa0 _ZN5dfcct20transform_goto_modelEv + 56
12  goto-instrument                     0x0000000100796e78 _ZN5dfcctC2ERK8optionstR11goto_modeltRK8dstringtRKN6nonstd13optional_lite8optionalINSt3__14pairIS5_S5_EEEEbRKNSB_3mapIS5_S5_NSB_4lessIS5_EENSB_9allocatorINSC_IS6_S5_EEEEEE24dfcc_loop_contract_modetR16message_handlertRKNSB_3setINSB_12basic_stringIcNSB_11char_traitsIcEENSK_IcEEEENSI_ISY_EENSK_ISY_EEEE + 404
13  goto-instrument                     0x0000000100796c50 _Z4dfccRK8optionstR11goto_modeltRK8dstringtRKN6nonstd13optional_lite8optionalINSt3__14pairIS4_S4_EEEEbRKNSA_3mapIS4_S4_NSA_4lessIS4_EENSA_9allocatorINSB_IS5_S4_EEEEEEbbRKNSA_3setINSA_12basic_stringIcNSA_11char_traitsIcEENSJ_IcEEEENSH_ISU_EENSJ_ISU_EEEER16message_handlert + 112
14  goto-instrument                     0x0000000100796564 _Z4dfccRK8optionstR11goto_modeltRK8dstringtRKN6nonstd13optional_lite8optionalIS4_EEbRKNSt3__13setIS4_NSD_4lessIS4_EENSD_9allocatorIS4_EEEEbbRKNSE_INSD_12basic_stringIcNSD_11char_traitsIcEENSH_IcEEEENSF_ISQ_EENSH_ISQ_EEEER16message_handlert + 304
15  goto-instrument                     0x00000001007ffbc4 _ZN30goto_instrument_parse_optionst23instrument_goto_programEv + 5092
16  goto-instrument                     0x00000001007fc0dc _ZN30goto_instrument_parse_optionst4doitEv + 552
17  goto-instrument                     0x0000000100bc43e8 _ZN19parse_options_baset4mainEv + 180
18  goto-instrument                     0x000000010076999c main + 48
19  dyld                                0x0000000194847f28 start + 2236


--- end invariant violation report ---
make: *** [ar_contracts.goto] Abort trap: 6
```

Any ideas where I went wrong?

--- a/regression/contracts-dfcc/loop_assigns_function_paramters/main.c
+++ b/regression/contracts-dfcc/loop_assigns_function_paramters/main.c
@@ -1,0 +1,19 @@
+#include <stdlib.h>
+
+void decr(size_t n)
+{
+  for(; n--;)
+    // clang-format off
+    __CPROVER_assigns(n)
+    __CPROVER_decreases(n)
+    // clang-format on
+    {
+    }
+}
+
+int main()
+{
+  size_t n;
+  decr(n);
+  return 0;
+}

--- a/regression/contracts-dfcc/loop_assigns_function_paramters/test.desc
+++ b/regression/contracts-dfcc/loop_assigns_function_paramters/test.desc
@@ -1,0 +1,11 @@
+CORE dfcc-only
+main.c
+--dfcc main --apply-loop-contracts --enforce-contract decr
+^\[decr.loop_assigns.\d+] line \d+ Check assigns clause inclusion for loop decr.0: SUCCESS$
+^VERIFICATION SUCCESSFUL$
+^EXIT=0$
+^SIGNAL=0$
+--
+--
+This test checks that function parameters are automatically added to the function
+write set when they are assigned from the body of a loop.

--- a/src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h
+++ b/src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h
@@ -291,6 +291,13 @@ public:
     return top_level_write_set;
   }
 
+  /// Returns the set of top level symbols that must be tracked explicitly in
+  /// the top level write set of the function.
+  const std::unordered_set<irep_idt> &get_top_level_tracked()
+  {
+    return top_level_tracked;
+  }
+
 private:
   const irep_idt &function_id;
   goto_functiont &goto_function;

--- a/src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp
+++ b/src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp
@@ -490,6 +490,25 @@ void dfcc_instrumentt::instrument_goto_function(
     insert_record_dead_call(function_id, write_set, local_static, end, body);
   }
 
+  const code_typet &code_type = to_code_type(
+    dfcc_utilst::get_function_symbol(goto_model.symbol_table, function_id)
+      .type);
+  const auto &top_level_tracked = cfg_info.get_top_level_tracked();
+
+  // automatically add/remove function parameters that must be tracked in the
+  // function write set (they must be explicitly tracked if they are assigned
+  // in the body of a loop)
+  for(const auto &param : code_type.parameters())
+  {
+    const irep_idt &param_id = param.get_identifier();
+    if(top_level_tracked.find(param_id) != top_level_tracked.end())
+    {
+      symbol_exprt param_symbol{param.get_identifier(), param.type()};
+      insert_add_decl_call(function_id, write_set, param_symbol, begin, body);
+      insert_record_dead_call(function_id, write_set, param_symbol, end, body);
+    }
+  }
+
   remove_skip(body);
 
   // recalculate numbers, etc.

